txpipe/magnification.py

```diff
@@ -0,0 +1,204 @@
+        "n_patches":20,
+        "bootstrap_error":True,
+    }
+
+    def run(self):
+        """
+        Run the analysis for this stage.
+        """
+        from scipy.stats import bootstrap
+
+        # load the catalogs
+        nomag_cat = self.open_input("binned_lens_catalog_nomag")
+        mag_cat   = self.open_input("binned_lens_catalog_mag")
+
+        # get/estimate the magnification applied to the magnified catalog
+        # TO DO: could put this as optional config item in TXSSIIngest
+        # so that it is done automatically
+        mu = self.config["applied_magnification"]
+        deltak = (1. - 1./mu)/2.
+        
+        if self.config["bootstrap_error"]:
+            #split no-mag sample up into patches to be used in the bootstrap errors
+            cluster = self.calc_cluster_patches(nomag_cat)
+
+        # compute fractional change in number count for both samples
+        # in each lens bin
+        nbins = nomag_cat['lens/'].attrs['nbin_lens']
+        outfile = self.setup_output(nbins)
+        for ibin in range(nbins):
+            print(f'Computing magnification coefficient for bin {ibin+1}')
+
+            # single redshift bins should be low enough number 
+            # count that we can load the whole sample
+            w1 = nomag_cat[f'lens/bin_{ibin}/weight'][:]
+            w2 = mag_cat[f'lens/bin_{ibin}/weight'][:]
+
+            #compute mag coeff with the whole sample
+            csample = self.calc_frac_change(w1, w2)/deltak
+            self.write_output(outfile, csample, ibin)
+
+            if self.config["bootstrap_error"]:
+                print(f'Computing uncertainty with bootstrap')
+
+                #compute mag coeff with subsampling patches to get a boostrap error
+                patch1 = cluster.predict(
+                    np.transpose(
+                        [ nomag_cat[f'lens/bin_{ibin}/ra'][:],
+                          nomag_cat[f'lens/bin_{ibin}/dec'][:], ]
+                        )
+                    )
+                patch2 = cluster.predict(
+                    np.transpose(
+                        [ mag_cat[f'lens/bin_{ibin}/ra'][:],
+                          mag_cat[f'lens/bin_{ibin}/dec'][:], ]
+                        )
+                    )
+
+                #count number of objects in each patch
+                label1, counts1 = np.unique(patch1, return_counts=True)
+                label2, counts2 = np.unique(patch2, return_counts=True)
+                assert (label1 == np.arange(cluster.n_clusters)).all(), "empty bootstrap patches"
+                assert (label2 == np.arange(cluster.n_clusters)).all(), "empty bootstrap patches"
+
+                # define a function that computes fractional number count change
+                # and takes only patch IDs as input (i.e. fix the counts to this z bin)
+                def calc_frac_change_patches_ibin(patch_ids):
+                    return self.calc_frac_change_patches(patch_ids, counts1, counts2)
```

Comment:
Is there a reason that the bootstrap part doesn't use the weights but the main calculation does?

Reply:
good catch. I've updated these to weighted counts